allennlp/data/token_indexers/pretrained_transformer_indexer.py

```diff
@@ -29,27 +31,50 @@ class PretrainedTransformerIndexer(TokenIndexer):
         We use a somewhat confusing default value of ``tags`` so that we do not add padding or UNK
         tokens to this namespace, which would break on loading because we wouldn't find our default
         OOV token.
+    num_added_start_tokens: ``int``, optional (default=1)
+        The number of start tokens that the tokenizer adds to a sequence.
+    num_added_end_tokens: ``int``, optional (default=1)
+        The number of end tokens that the tokenizer adds to a sequence.
+    intra_word_tokenization: ``bool``, optional (default=False)
+        If True, further tokenize each token into subword tokens using the corresponding tokenizer.
+        Should be set to the same value as the ``intra_word_tokenized`` option on the
+        :class:`PretrainedTransformerEmbedder`.
     """
 
     def __init__(
-        self, model_name: str, namespace: str = "tags", token_min_padding_length: int = 0
+        self,
+        model_name: str,
+        namespace: str = "tags",
+        num_added_start_tokens: int = 1,
```

Comment:
Earlier you mentioned that you could auto-detect these, but since they were needed in the embedder we decided to specify them explicitly. However, I think you managed to avoid that in this version of the code. Is it possible to leave this out now?

Reply:
The embedder will need those variables when I submit the long sequence wrapping PR. It'd be a little troublesome to add back that logic now and remove it in that PR. With that said, if we think the auto-detection logic is better, we can always add that back, perhaps as a standalone PR, to replace config options.

Reply:
The interface between the indexer and the embedder can be arbitrary - you can pass whatever you need.  Based on the `assert` below, it looks like you can compute this automatically.  I would definitely say you should remove these as parameters here.

We sometimes have things like this be parameters on the _modeling_ side, where we need to know an embedding dimension up front, for example.  So that would mean having a parameter for it in the `Embedder`.  But there's no reason to have it as a parameter here, even if you need it in the `Embedder`.  But even better than making it a parameter in the `Embedder`, you can just compute it here and pass it through in your tensor dictionary, and the embedder can get the right value that way.

Reply:
@matt-gardner, wouldn't we need to pass along a full batch of these? I suppose that's likely not significant in terms memory, but it does seem a bit hacky. Maybe something we can use to inform future API improvements.

Reply:
You can make the interface between the indexer and the embedder whatever you want, including controlling [how things get batched](https://github.com/allenai/allennlp/blob/52bece613672d8b289b722ddc86731ded07cca3f/allennlp/data/token_indexers/token_indexer.py#L89).  You can just not batch this field.

Reply:
Hmm, I was too hasty, that just does padding, not batching.  But, I think the additional overhead is pretty minor here, relative to requiring a user to figure out the specifics of the what the underlying tokenizer is doing just to use the new model.

Reply:
Huggingface gives us `num_added_tokens` which a sum of my `num_added_start_tokens` and `num_added_end_tokens`. Those values are not provided. The `assert` only acts as a checksum. Yes there is a way (read: hack) I can auto-determine these two values by feeding a dummy sentence to the tokenizer. However, both the indexer and embedder need to know these two values (not right now, but in a next PR coming very soon). I was not aware of a simple way to pass this information across, and I really don't want to repeat this hack in two different places.

Reply:
If we really care, we could allow the indexer to control batching itself; that would be a non-user-facing change that would give us a very slight gain in efficiency.  The `TextField` would just delegate batching to the underlying indexers.  I don't really think that's worth it, but we could if we wanted.

Reply:
Ok, I will add back the auto-determination logic back in, and then pass that information to the embedder via `tokens_to_indices` output. On a related note, in the future if similar issues come up again, it might be nice to give the embedder a reference to the indexer in its `__init__`.

Reply:
In the original design of the precursor to this library, the indexer and the embedder were a single object.  There clearly is a strong relationship between these two, and having them share state can be helpful in some cases, like this one.  We decided, though, that decoupling the data processing from the model is more important and in the end cleaner overall.

On the auto-determination logic, it's probably worth at least a little thought on if this can be added upstream in the huggingface library, so that we can just call methods there that give us what we need.

Reply:
Sounds good. I will first add that as a function in AllenNLP, and will submit a PR to huggingface if I got a chance.